README.md
# htknx

[![PyPI version](https://img.shields.io/pypi/v/htknx.svg)](https://pypi.org/project/htknx)
[![Python versions](https://img.shields.io/pypi/pyversions/htknx.svg)](https://pypi.org/project/htknx)
[![License](https://img.shields.io/pypi/l/htknx.svg)](https://pypi.org/project/htknx)
[![Build status](https://github.com/dstrigl/HtREST/workflows/CI/badge.svg)](https://github.com/dstrigl/htknx/actions?query=workflow%3ACI)
[![Updates](https://pyup.io/repos/github/dstrigl/htknx/shield.svg)](https://pyup.io/repos/github/dstrigl/htknx)

![KNX](knx-logo.png)

[Heliotherm](http://www.heliotherm.com/) heat pump [KNX](https://www.knx.org) gateway for Python 3.7 and 3.8.

Can be used to provide the different [Heliotherm heat pump parameters](https://htheatpump.readthedocs.io/en/latest/htparams.html) as data points to the KNX bus.


* GitHub repo: https://github.com/dstrigl/htknx
* Free software: [GNU General Public License v3](https://www.gnu.org/licenses/gpl-3.0.en.html)


> **Warning:**
>
> Please note that any incorrect or careless usage of this application as well as
> errors in the implementation can damage your heat pump!
>
> Therefore, the author does not provide any guarantee or warranty concerning
> to correctness, functionality or performance and does not accept any liability
> for damage caused by this application, examples or mentioned information.
>
> **Thus, use it on your own risk!**


### Wanna support me?

[![Buy Me A Coffee](https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png)](https://www.buymeacoffee.com/N362PLZ)


## Setup

The following diagram shows the schematic view of a possible setup:

![Setup](setup.png)


## Installation

You can install or upgrade `htknx` with:

```
$ pip install htknx --upgrade
```

Or you can install from source with:

```
$ git clone https://github.com/dstrigl/htknx.git
$ cd htknx
$ python setup.py install
```


## Usage

```
usage: htknx [-h] [--logging-config LOGGING_CONFIG] [config_file]

Heliotherm heat pump KNX gateway, v0.1.0.

  https://github.com/dstrigl/htknx

positional arguments:
  config_file           the filename under which the gateway settings can be
                        found, default: htknx.yaml

optional arguments:
  -h, --help            show this help message and exit
  --logging-config LOGGING_CONFIG
                        the filename under which the logging configuration can
                        be found, default: logging.conf

DISCLAIMER
----------
  Please note that any incorrect or careless usage of this program as well as
  errors in the implementation can damage your heat pump!
  Therefore, the author does not provide any guarantee or warranty concerning
  to correctness, functionality or performance and does not accept any liability
  for damage caused by this program or mentioned information.
  Thus, use it on your own risk!

```


### Example:

```
$ htknx /home/pi/my-htknx.yaml
HTHEATPUMP: load parameter definitions from: /home/pi/venv/htknx/lib/python3.7/site-packages/htheatpump-1.3.1-py3.7.egg/htheatpump/htparams.csv
Namespace(config_file='/home/pi/my-htknx.yaml', logging_config='/home/pi/prog/htknx/htknx/logging.conf')
2021-01-27 17:14:35,736 INFO [htknx.__main__|main_async]: Load settings from '/home/pi/my-htknx.yaml'.
2021-01-27 17:14:36,394 INFO [htknx.__main__|main_async]: Start Heliotherm heat pump KNX gateway v0.1.0.
2021-01-27 17:14:36,505 INFO [htknx.__main__|main_async]: Connected successfully to heat pump with serial number 123456.
2021-01-27 17:14:36,551 INFO [htknx.__main__|main_async]: Software version = 3.0.20 (2321)
2021-01-27 17:14:36,587 WARNING [xknx.log|loop_until_sigint]: Press Ctrl+C to stop
2021-01-27 17:14:36,592 INFO [htknx.__main__|login_loop]: <<< [ LOGIN (every 0:00:30) ] >>>
2021-01-27 17:14:36,598 INFO [htknx.__main__|update_loop]: <<< [ UPDATE (every 0:00:25) ] >>>
2021-01-27 17:14:36,604 INFO [htknx.__main__|cyclic_sending_loop]: <<< [ CYCLIC SENDING (every 0:01:00) ] >>>
2021-01-27 17:14:36,608 INFO [htknx.__main__|cyclic_sending_loop]: Sending: ['Temp. Aussen', 'Stoerung']
2021-01-27 17:14:37,152 INFO [htknx.__main__|update_loop]: Update: {'Betriebsart': 1, 'HKR Soll_Raum': 22.0, 'WW Normaltemp.': 50, 'BSZ Verdichter Betriebsst. ges': 12000, 'Temp. Aussen': -2.6, 'Temp. Frischwasser_Istwert': 43.6, 'Heizkreispumpe': True, 'Stoerung': False}
2021-01-27 17:14:37,156 INFO [htknx.htdatapoint|set]: Update and send DP 'Betriebsart' [1/7/9]: value=1 (send_on_change: True, on_change_of_absolute: 1, on_change_of_relative: None, last_sent_value: None)
2021-01-27 17:14:37,161 INFO [htknx.htdatapoint|set]: Update and send DP 'HKR Soll_Raum' [1/7/10]: value=22.0 (send_on_change: True, on_change_of_absolute: 0.1, on_change_of_relative: None, last_sent_value: None)
2021-01-27 17:14:37,165 INFO [htknx.htdatapoint|set]: Update and send DP 'WW Normaltemp.' [1/7/17]: value=50 (send_on_change: True, on_change_of_absolute: 1, on_change_of_relative: None, last_sent_value: None)
2021-01-27 17:14:37,170 INFO [htknx.htdatapoint|set]: Update and send DP 'Temp. Frischwasser_Istwert' [1/7/45]: value=43.6 (send_on_change: True, on_change_of_absolute: None, on_change_of_relative: 10, last_sent_value: None)
2021-01-27 17:15:02,181 INFO [htknx.__main__|update_loop]: <<< [ UPDATE (every 0:00:25) ] >>>
2021-01-27 17:15:02,633 INFO [htknx.__main__|update_loop]: Update: {'Betriebsart': 1, 'HKR Soll_Raum': 22.0, 'WW Normaltemp.': 50, 'BSZ Verdichter Betriebsst. ges': 12001, 'Temp. Aussen': -2.6, 'Temp. Frischwasser_Istwert': 43.6, 'Heizkreispumpe': True, 'Stoerung': False}
2021-01-27 17:15:06,641 INFO [htknx.__main__|login_loop]: <<< [ LOGIN (every 0:00:30) ] >>>
2021-01-27 17:15:27,662 INFO [htknx.__main__|update_loop]: <<< [ UPDATE (every 0:00:25) ] >>>
2021-01-27 17:15:28,107 INFO [htknx.__main__|update_loop]: Update: {'Betriebsart': 1, 'HKR Soll_Raum': 22.0, 'WW Normaltemp.': 50, 'BSZ Verdichter Betriebsst. ges': 12001, 'Temp. Aussen': -2.6, 'Temp. Frischwasser_Istwert': 43.6, 'Heizkreispumpe': True, 'Stoerung': False}
2021-01-27 17:15:36,624 INFO [htknx.__main__|cyclic_sending_loop]: <<< [ CYCLIC SENDING (every 0:01:00) ] >>>
2021-01-27 17:15:36,629 INFO [htknx.__main__|cyclic_sending_loop]: Sending: ['Temp. Aussen', 'Stoerung']
2021-01-27 17:15:36,688 INFO [htknx.__main__|login_loop]: <<< [ LOGIN (every 0:00:30) ] >>>
...
```


## Configuration

`htknx` is controlled via a configuration file. Per default the configuration file is named `htknx.yaml`.

The configuration file can contain the following five sections:

* The `general` section can contain:

    * `update_interval` the update interval to refresh the heat pump parameters (optional, default: `60` seconds)
    * `cyclic_sending_interval` the time interval for data points that are to be sent cyclically to the KNX bus (optional, default: `10` minutes)
    * `synchronize_clock_weekly` to define the day of the week (`mon`, `tue`, `wed`, `thu`, `fri`, `sat`, `sun`) and the time (e.g. `'19:55'`) to synchronize the heat pump clock regularly (optional, default: disabled)

* The `heat_pump` section is needed to specify the connection to the heat pump:

    * `device` the serial device on which the heat pump is connected (e.g. `/dev/ttyUSB0`)
    * `baudrate` baudrate of the serial connection to the heat pump (same as configured on the heat pump, e.g. `19200`)

* The `knx` section is needed to specify the connection to the KNX interface (e.g. a [Weinzierl KNX IP Interface 731](https://www.weinzierl.de/index.php/de/alles-knx1/knx-devices/knx-ip-interface-731-de)):

    * `gateway_ip` the ip address of the KNX tunneling interface (e.g. `192.168.11.81`)
    * `gateway_port` the port the KNX tunneling interface is listening on (optional, default: `3671`)
    * `auto_reconnect` determines whether to try a reconnect if the connection to the KNX tunneling interface could not be established (optional, default: `true`)
    * `auto_reconnect_wait` the time to wait for the next auto reconnect (optional, default: `3` seconds)
    * `local_ip` the local ip address that is used to connect to the KNX tunneling interface (optional, e.g. `192.168.11.114`)
    * `own_address` the individual (physical) address of this gateway (optional, default: `15.15.250`)
    * `rate_limit` a rate limit for telegrams sent to the KNX bus per second (optional, default: `10`)

* The `data_points` section contains the dictionary of [heat pump parameters](https://htheatpump.readthedocs.io/en/latest/htparams.html) for which a data point should be provided to the KNX bus.

  Each item in the dictionary consists of the "parameter name" as key and the following properties:

    * `value_type` the value type of the data point (e.g. `binary`, `common_temperature`, `1byte_unsigned`, `4byte_unsigned`, etc. as supported by [XKNX](https://github.com/XKNX/xknx))
    * `group_address` the KNX group address of the data point (e.g. `1/2/3`)
    * `writable` determines whether the data point could also be written or not (optional, default: `false`)
    * `cyclic_sending` determines whether the data point should be sent cyclically to the KNX bus (optional, default: `false`)
    * `send_on_change` defines whether the data point should be sent to the KNX bus if it changes for a defined value (optional, default: `false`)
    * `on_change_of_absolute` the absolute value of change for sending on change (e.g. `0.5` for 0.5°C)
    * `on_change_of_relative` the relative value of change for sending on change (in percent, e.g. `10` for 10%)

  A list of supported value types can be found in the comments of the [configuration template](https://github.com/dstrigl/htknx/blob/master/htknx/htknx-template.yaml) or [sample configuration file](https://github.com/dstrigl/htknx/blob/master/htknx/htknx.yaml). These are exactly the same value types supported by the [XKNX](https://github.com/XKNX/xknx) module on which this project is based.

* The `notifications` section contains the setup of the different supported notifications (optional).

  At the moment the following notifications are supported:

  * `on_malfunction` which sends a notification with the current error message (as DPT-16.000) to the KNX bus if the heat pump is malfunctioning:

      * `group_address` the KNX group address under which the error message is sent (e.g. `1/2/255`)
      * `repeat_after` the time interval until the notification should be repeated if the heat pump is still malfunctioning (optional, e.g.  `10` minutes)


### Sample configuration:

The following block shows a sample configuration for the heat pump parameters

  * *Betriebsart*,
  * *HKR Soll_Raum*,
  * *WW Normaltemp.*,
  * *BSZ Verdichter Betriebsst. ges*,
  * *Temp. Aussen*,
  * *Temp. Frischwasser_Istwert*,
  * *Heizkreispumpe* and
  * *Stoerung*.

These heat pump parameters are updated every 25 seconds and some of them are sent cyclically to the KNX bus every minute, while some of the other parameters are sent immediately after a change.

Only the three parameters *"Betriebsart"*, *"HKR Soll_Raum"* and *"WW Normaltemp."* are writable.

In addition, a notification with the current error message is sent to the KNX bus if the heat pump is malfunctioning.

```
general:
  update_interval:
    seconds: 25
  cyclic_sending_interval:
    minutes: 1
#  synchronize_clock_weekly:
#    weekday: tue  # mon, tue, wed, thu, fri, sat, sun
#    time: '13:05'

heat_pump:
  device: /dev/ttyUSB0
  baudrate: 115200

knx:
  gateway_ip: '192.168.11.81'
  rate_limit: 10
#  gateway_port: 3671
#  auto_reconnect: True
#  auto_reconnect_wait:
#    seconds: 3
#  local_ip: '192.168.11.140'
#  own_address: '15.15.250'

data_points:
  # https://htheatpump.readthedocs.io/en/latest/htparams.html
  # -----------------------------------------------------------------
  'Betriebsart':
    # 0 = Aus
    # 1 = Automatik
    # 2 = Kühlen
    # 3 = Sommer
    # 4 = Dauerbetrieb
    # 5 = Absenkung
    # 6 = Urlaub
    # 7 = Party
    value_type: '1byte_unsigned'  # DPT-5.005 (Dezimalfaktor 0..255)
    group_address: '1/7/9'
    writable: true
    send_on_change: true
    on_change_of_absolute: 1
  # -----------------------------------------------------------------
  'HKR Soll_Raum':
    # MIN: 10.0
    # MAX: 25.0
    value_type: 'common_temperature'  # DPT-14.068 (Temperatur °C)
    group_address: '1/7/10'
    writable: true
    send_on_change: true
    on_change_of_absolute: 0.1
  # -----------------------------------------------------------------
  'WW Normaltemp.':
    # MIN: 10
    # MAX: 75
    value_type: '1byte_unsigned'  # DPT-5.005 (Dezimalfaktor 0..255)
    # or:
    # value_type: 'common_temperature'  # DPT-14.068 (Temperatur °C)
    group_address: '1/7/17'
    writable: true
    send_on_change: true
    on_change_of_absolute: 1
  # -----------------------------------------------------------------
  'BSZ Verdichter Betriebsst. ges':
    # MIN: 0
    # MAX: 100000
    value_type: '4byte_unsigned'  # DPT-12.102 (Zähler Zeit Stunden)
    group_address: '1/7/31'
  # -----------------------------------------------------------------
  'Temp. Aussen':
    # MIN: -20.0
    # MAX: 40.0
    value_type: 'common_temperature'  # DPT-14.068 (Temperatur °C)
    group_address: '1/7/36'
    cyclic_sending: true
  # -----------------------------------------------------------------
  'Temp. Frischwasser_Istwert':
    # MIN: 0.0
    # MAX: 70.0
    value_type: 'common_temperature'  # DPT-14.068 (Temperatur °C)
    group_address: '1/7/45'
    send_on_change: true
    on_change_of_relative: 10
  # -----------------------------------------------------------------
  'Heizkreispumpe':
    # MIN: 0
    # MAX: 1
    value_type: 'binary'  # DPT-1.011 (Status)
    group_address: '1/7/51'
    send_on_change: true
  # -----------------------------------------------------------------
  'Stoerung':
    # MIN: 0
    # MAX: 1
    value_type: 'binary'  # DPT-1.005 (Alarm)
    group_address: '1/7/56'
    cyclic_sending: true

notifications:
  on_malfunction:
    group_address: '1/7/255'  # DPT-16.000 (Zeichen ASCII)
    repeat_after:
      minutes: 10
```

The following image shows a screenshot of the [Group Monitor](https://support.knx.org/hc/en-us/articles/360011717719-Monitor) of the [ETS application](https://www.knx.org/knx-en/for-professionals/software/ets-5-professional/) with the data points provided by the upper [sample configuration](https://github.com/dstrigl/htknx/blob/master/htknx/htknx.yaml):

![ETS Group Monitor](group-monitoring.png)


## Credits

* [XKNX](https://xknx.io/) - Asynchronous Python Library for KNX
* Project dependencies scanned by [PyUp.io](https://pyup.io)


## License

Distributed under the terms of the [GNU General Public License v3](https://www.gnu.org/licenses/gpl-3.0.en.html).


HISTORY.md
# History

## 0.2.0 (2021-02-06)

* Added functionality to synchronize the clock of the heat pump regularly (see `synchronize_clock_weekly` in section `general`).

## 0.1.0 (2021-01-30)

* First release on PyPI.


htknx/config.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#  htknx - Heliotherm heat pump KNX gateway
#  Copyright (C) 2021  Daniel Strigl

#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

""" Parsing a given config file in YAML format. """

import logging
from datetime import timedelta
from typing import Any, Callable, Dict

import voluptuous as vol
import yaml
from htheatpump.htparams import HtParams
from xknx import XKNX
from xknx.io import ConnectionConfig, ConnectionType
from xknx.telegram import IndividualAddress

from . import config_validation as cv

_LOGGER = logging.getLogger(__name__)


CONF_GENERAL = "general"
CONF_UPDATE_INTERVAL = "update_interval"
CONF_CYCLIC_SENDING_INTERVAL = "cyclic_sending_interval"
CONF_SYNCHRONIZE_CLOCK_WEEKLY = "synchronize_clock_weekly"
CONF_SYNCHRONIZE_CLOCK_WEEKDAY = "weekday"
CONF_SYNCHRONIZE_CLOCK_TIME = "time"

CONF_HEAT_PUMP = "heat_pump"
CONF_DEVICE = "device"
CONF_BAUDRATE = "baudrate"

CONF_KNX = "knx"
CONF_GATEWAY_IP = "gateway_ip"
CONF_GATEWAY_PORT = "gateway_port"
CONF_AUTO_RECONNECT = "auto_reconnect"
CONF_AUTO_RECONNECT_WAIT = "auto_reconnect_wait"
CONF_LOCAL_IP = "local_ip"
CONF_OWN_ADDRESS = "own_address"
CONF_RATE_LIMIT = "rate_limit"

CONF_DATA_POINTS = "data_points"
CONF_VALUE_TYPE = "value_type"
CONF_GROUP_ADDRESS = "group_address"
CONF_WRITABLE = "writable"
CONF_CYCLIC_SENDING = "cyclic_sending"
CONF_SEND_ON_CHANGE = "send_on_change"
CONF_ON_CHANGE_OF_ABSOLUTE = "on_change_of_absolute"
CONF_ON_CHANGE_OF_RELATIVE = "on_change_of_relative"

CONF_NOTIFICATIONS = "notifications"
CONF_ON_MALFUNCTION = "on_malfunction"
CONF_REPEAT_AFTER = "repeat_after"


DEFAULT_UPDATE_INTERVAL = 60
DEFAULT_CYCLIC_SENDING_INTERVAL = 600
DEFAULT_BAUDRATE = 115200
DEFAULT_GATEWAY_PORT = 3671
DEFAULT_AUTO_RECONNECT_WAIT = 3
DEFAULT_RATE_LIMIT = 10  # XKNX.DEFAULT_RATE_LIMIT


SYNCHRONIZE_CLOCK_WEEKLY_SCHEMA = vol.Schema(
    {
        CONF_SYNCHRONIZE_CLOCK_WEEKDAY: vol.All(cv.string, vol.In(cv.WEEKDAYS)),
        CONF_SYNCHRONIZE_CLOCK_TIME: cv.time,
    }
)

GENERAL_SCHEMA = vol.Schema(
    {
        vol.Optional(
            CONF_UPDATE_INTERVAL, default=DEFAULT_UPDATE_INTERVAL
        ): cv.time_interval,
        vol.Optional(
            CONF_CYCLIC_SENDING_INTERVAL, default=DEFAULT_CYCLIC_SENDING_INTERVAL
        ): cv.time_interval,
        vol.Optional(CONF_SYNCHRONIZE_CLOCK_WEEKLY): SYNCHRONIZE_CLOCK_WEEKLY_SCHEMA,
    }
)

HEAT_PUMP_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_DEVICE): cv.string,
        vol.Optional(CONF_BAUDRATE, default=DEFAULT_BAUDRATE): vol.All(
            vol.Coerce(int), vol.In([9600, 19200, 38400, 57600, 115200])
        ),
    }
)

KNX_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_GATEWAY_IP): cv.string,
        vol.Optional(CONF_GATEWAY_PORT, default=DEFAULT_GATEWAY_PORT): cv.port,
        vol.Optional(CONF_AUTO_RECONNECT, default=True): cv.boolean,
        vol.Optional(
            CONF_AUTO_RECONNECT_WAIT, default=DEFAULT_AUTO_RECONNECT_WAIT
        ): cv.time_interval,
        vol.Optional(CONF_LOCAL_IP): cv.string,
        vol.Optional(
            CONF_OWN_ADDRESS, default=XKNX.DEFAULT_ADDRESS
        ): cv.ensure_individual_address,
        vol.Optional(CONF_RATE_LIMIT, default=DEFAULT_RATE_LIMIT): vol.All(
            vol.Coerce(int), vol.Range(min=0, max=100)
        ),
    }
)


def validate_data_point() -> Callable:
    """Ensure that the data point is valid."""

    def validate(obj: Dict) -> Dict:
        if (
            obj[CONF_VALUE_TYPE] == "binary"
            and len({CONF_ON_CHANGE_OF_ABSOLUTE, CONF_ON_CHANGE_OF_RELATIVE} & set(obj))
            > 0
        ):
            raise vol.Invalid(
                "{} not allowed for binary data point".format(
                    ", ".join((CONF_ON_CHANGE_OF_ABSOLUTE, CONF_ON_CHANGE_OF_RELATIVE))
                )
            )
        if (
            obj[CONF_VALUE_TYPE] != "binary"
            and obj[CONF_SEND_ON_CHANGE]
            and len({CONF_ON_CHANGE_OF_ABSOLUTE, CONF_ON_CHANGE_OF_RELATIVE} & set(obj))
            < 1
        ):
            raise vol.Invalid(
                "must contain at least one of {}".format(
                    ", ".join((CONF_ON_CHANGE_OF_ABSOLUTE, CONF_ON_CHANGE_OF_RELATIVE))
                )
            )

        return obj

    return validate


DATA_POINT_SCHEMA = vol.All(
    dict,
    vol.Schema(
        {
            vol.Required(CONF_VALUE_TYPE): vol.Or(cv.ensure_knx_dpt, "binary"),
            vol.Required(CONF_GROUP_ADDRESS): cv.ensure_group_address,
            vol.Optional(CONF_WRITABLE, default=False): cv.boolean,
            vol.Optional(CONF_CYCLIC_SENDING, default=False): cv.boolean,
            vol.Optional(CONF_SEND_ON_CHANGE, default=False): cv.boolean,
            vol.Exclusive(
                CONF_ON_CHANGE_OF_ABSOLUTE,
                "on_change_of",
                msg="absolute or relative change",
            ): cv.number_greater_zero,
            vol.Exclusive(
                CONF_ON_CHANGE_OF_RELATIVE,
                "on_change_of",
                msg="absolute or relative change",
            ): cv.number_greater_zero,
        }
    ),
    validate_data_point(),
)


def check_for_valid_parameter_names() -> Callable:
    """Check for valid parameter names in the data points section."""

    def validate(obj: Dict) -> Dict:
        for name in obj.keys():
            if name not in HtParams.keys():
                raise vol.Invalid(f"{name!r} is not a valid heat pump parameter")

        return obj

    return validate


def check_for_warnings_in_data_points() -> Callable:
    """Check for warnings in the data point config section."""

    def validate(obj: Dict) -> Dict:
        for name, dp in obj.items():
            if (
                dp[CONF_VALUE_TYPE] != "binary"
                and not dp[CONF_SEND_ON_CHANGE]
                and len(
                    {CONF_ON_CHANGE_OF_ABSOLUTE, CONF_ON_CHANGE_OF_RELATIVE} & set(dp)
                )
                > 0
            ):
                _LOGGER.warning(
                    "%s is defined, but %s is set to false for data point '%s'",
                    CONF_ON_CHANGE_OF_ABSOLUTE
                    if CONF_ON_CHANGE_OF_ABSOLUTE in dp
                    else CONF_ON_CHANGE_OF_RELATIVE,
                    CONF_SEND_ON_CHANGE,
                    name,
                )

        return obj

    return validate


DATA_POINTS_SCHEMA = vol.All(
    dict,
    vol.Schema({cv.string: DATA_POINT_SCHEMA}),
    check_for_valid_parameter_names(),
    check_for_warnings_in_data_points(),
)

NOTIFICATION_SCHEMA = vol.Schema(
    {vol.Required(CONF_GROUP_ADDRESS): cv.ensure_group_address}
)

ON_MALFUNCTION_SCHEMA = NOTIFICATION_SCHEMA.extend(
    {
        vol.Optional(CONF_REPEAT_AFTER, default=None): vol.Or(cv.time_interval, None),
    }
)

NOTIFICATIONS_SCHEMA = vol.Schema({CONF_ON_MALFUNCTION: ON_MALFUNCTION_SCHEMA})

CONFIG_SCHEMA = vol.Schema(
    {
        vol.Optional(CONF_GENERAL): GENERAL_SCHEMA,
        CONF_HEAT_PUMP: HEAT_PUMP_SCHEMA,
        CONF_KNX: KNX_SCHEMA,
        vol.Optional(CONF_DATA_POINTS): DATA_POINTS_SCHEMA,
        vol.Optional(CONF_NOTIFICATIONS): NOTIFICATIONS_SCHEMA,
    }
)


class Config:
    """Class for parsing a given config file, e.g. 'htknx.yaml'."""

    def __init__(self) -> None:
        """Initialize the Config class."""
        self.general: Dict[str, Any] = {
            CONF_UPDATE_INTERVAL: timedelta(DEFAULT_UPDATE_INTERVAL),
            CONF_CYCLIC_SENDING_INTERVAL: timedelta(DEFAULT_CYCLIC_SENDING_INTERVAL),
            CONF_SYNCHRONIZE_CLOCK_WEEKLY: None,
        }
        self.heat_pump: Dict[str, Any] = {
            CONF_DEVICE: None,
            CONF_BAUDRATE: DEFAULT_BAUDRATE,
        }
        self.knx: Dict[str, Any] = {
            "connection_config": ConnectionConfig(
                connection_type=ConnectionType.TUNNELING
            ),
            CONF_OWN_ADDRESS: IndividualAddress(XKNX.DEFAULT_ADDRESS),
            CONF_RATE_LIMIT: DEFAULT_RATE_LIMIT,
        }
        self.data_points: Dict[str, dict] = {}
        self.notifications: Dict[str, dict] = {}

    def read(self, filename: str = "htknx.yaml") -> None:
        """Read the configuration from the given file.

        :param filename: The filename to read the configuration from, e.g. 'htknx.yaml'.
        :type filename: str
        """
        # _LOGGER.debug("Reading config file '%s'.", filename)
        with open(filename, "r") as f:
            doc = yaml.safe_load(f)
            doc = CONFIG_SCHEMA(doc)
            self._parse_general_settings(doc)
            self._parse_heat_pump_settings(doc)
            self._parse_knx_settings(doc)
            self._parse_data_points(doc)
            self._parse_notifications(doc)

    def _parse_general_settings(self, doc) -> None:
        """Parse the general section of the config file."""
        if CONF_GENERAL in doc:
            self.general.update(doc[CONF_GENERAL])

    def _parse_heat_pump_settings(self, doc) -> None:
        """Parse the heat pump section of the config file."""
        if CONF_HEAT_PUMP in doc:
            self.heat_pump.update(doc[CONF_HEAT_PUMP])

    def _parse_knx_settings(self, doc) -> None:
        """Parse the KNX section of the config file."""
        if CONF_KNX in doc:
            if CONF_GATEWAY_IP in doc[CONF_KNX]:
                self.knx["connection_config"].gateway_ip = doc[CONF_KNX][
                    CONF_GATEWAY_IP
                ]
            if CONF_GATEWAY_PORT in doc[CONF_KNX]:
                self.knx["connection_config"].gateway_port = doc[CONF_KNX][
                    CONF_GATEWAY_PORT
                ]
            if CONF_AUTO_RECONNECT in doc[CONF_KNX]:
                self.knx["connection_config"].auto_reconnect = doc[CONF_KNX][
                    CONF_AUTO_RECONNECT
                ]
            if CONF_AUTO_RECONNECT_WAIT in doc[CONF_KNX]:
                self.knx["connection_config"].auto_reconnect_wait = doc[CONF_KNX][
                    CONF_AUTO_RECONNECT_WAIT
                ].total_seconds()
            if CONF_LOCAL_IP in doc[CONF_KNX]:
                self.knx["connection_config"].local_ip = doc[CONF_KNX][CONF_LOCAL_IP]
            if CONF_OWN_ADDRESS in doc[CONF_KNX]:
                self.knx[CONF_OWN_ADDRESS] = IndividualAddress(
                    doc[CONF_KNX][CONF_OWN_ADDRESS]
                )
            if CONF_RATE_LIMIT in doc[CONF_KNX]:
                self.knx[CONF_RATE_LIMIT] = doc[CONF_KNX][CONF_RATE_LIMIT]

    def _parse_data_points(self, doc) -> None:
        """Parse the data points section of the config file."""
        if CONF_DATA_POINTS in doc:
            self.data_points.update(doc[CONF_DATA_POINTS])

    def _parse_notifications(self, doc) -> None:
        """Parse the notifications section of the config file."""
        if CONF_NOTIFICATIONS in doc:
            self.notifications.update(doc[CONF_NOTIFICATIONS])
